### DISCUSSION TITLE
TableView?

### DISCUSSION THREAD

**Discussion by jamesblacklock** [slint-ui/slint#574] [Q&A]
I would like to work on implementing a general purpose TableView, kind of like a ListView/GridLayout combination that would accept Rows and also have adjustable width columns. Ideally I'd like to contribute this to the project, so I'm looking for any feedback on this idea before I get started.

**Comment by jamesblacklock** [comment]
Also seems important to add support for `Rows` in `for` expressions, which I could also look at doing

**Reply by ogoffart** [reply]
Supporting `for` in `Rows` seems likea good feature that would help to simulate a TableView and that shouldn't be too difficult to implement.

The key is that the layout for the GridLayout is done by the `sixtyfps_corelib::layout::solve_grid_layout` and `sixtyfps_corelib::layout::grid_layout_info`.  But the sixtyfps compiler currently only support generating a static array of cells to pass to these function.  The place that generate this array for rust is `grid_layout_cell_data` https://github.com/sixtyfpsui/sixtyfps/blob/bb4e5d8b5501e74f506c6bcd77d7ee4c402a3de4/sixtyfps_compiler/generator/rust.rs#L1800  and there are similar function in cpp.rs and in eval_layout.rs.  Note that for example, the box_layout_cell_data supports dynamic array (the `repeater_count` by generating code that fills a vector), so this can be used as an inspiration to implement the same for the GridLayout

So what need to be done is
 1. change lower_layout.rs not to throw an error if there is a repeated element in a grid layout. https://github.com/sixtyfpsui/sixtyfps/blob/bb4e5d8b5501e74f506c6bcd77d7ee4c402a3de4/sixtyfps_compiler/passes/lower_layout.rs#L247   Note that this has implications on the way we compute the rows and column statically.  I guess we could accept it only if it is alone in a Row, so we assign the `row` and the column would be the index.
 2. adjust the generator in rust.rs, cpp.rs and  eval_layout.rs. to handle these repeated element and build the cells in a vector.

**Reply by ogoffart** [reply]
btw, the issue for this is #358

**Comment by ogoffart** [comment]
One way to implement the TableView would be to use a ListView and support a header.
That's a bit what the sysinfo example does:
https://github.com/sixtyfpsui/sixtyfps/blob/bb4e5d8b5501e74f506c6bcd77d7ee4c402a3de4/examples/bash/sysinfo.60#L140
It sizes each column of the row to the size of the header.  And one could imagine having the header be interactive and allow resizing.
But if we want to automatically size the column based on the content, we need somehow specific code that compute the size of each column.
So we need something in the native code like the `compute_layout_listview` but works on model specific for a TableView.

API wise, we need to find out the best API for both the .60 code, and the native code.
Do we want to have a special `Model` class for TableView?  Or is it a Model of a Model.

For example, something like (pseudo .60 code)
```
TableView := ListView {
   property <[ [StandardListViewItem] ]> model;  // a model of a model of StandardListViewItem
   for row in model: HorizontalLayout {
      for col in row: TableViewCell {
          // ... 
      }
   }
}
```

**Reply by jamesblacklock** [reply]
Yes, right now I am using a ListView with a header, but getting column width right is tricky. There needs to be some code running behind the scenes to make this work seamlessly, I think. Still figuring it out.

**Comment by aetelani** [comment]
My [approach](https://github.com/aetelani/slint-ui-testing/blob/f85f4fc7533d869a3344cd91e441fa82db867363/src/main.rs#L59) was to introduce second model and map indexes.  Problem was that the original model was iterated to end so there was empty rows which were counted to height and affected as well to scrollbar size.  The cell width needs to be constant as well to align nicely (I had also perf issue on Win without fixing length).
Naive fix could be just provide way to exit the _for_ iterator or have property _cell-amount_ to that would count height correctly
```
              sv := ListView {
                      for it[ind] in model:
                      rb := Rectangle {
// BUG: Cols does not really work as the table is over-indexed with empty text fields
// Added map-ind constraint but scroll bars are still too big
                              property<[int]> r-model: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
                              hb := HorizontalBox {
                              for r-ind in r-model:
                              rect := Rectangle {
                                  callback map-ind() -> int;
                                  map-ind() => { (ind * r-model.length) + r-ind }
...
```
